sets/slices.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sets
+
+// Intersect finds the intersection of all slices, where intersection is defined
+// as elements that exist in all slices. The elements in the returned slice will
+// be in an undefined order - use [IntersectStable] to trade efficiency for
+// ordering. Duplicate elements are removed. This function always returns an
+// allocated slice, even if the intersection is the empty set.
+//
+// It does not modify any of the given slices, but also does not deep copy any
+// values. That means the returned slice may have elements that point to the
+// same objects as in the original slice.
+func Intersect[T comparable](slices ...[]T) []T {
+	if len(slices) == 0 {
+		return []T{}
+	}
+
+	maps := make([]map[T]struct{}, len(slices))
+	for i, s := range slices {
+		// Short-circuit: if any of the slices are the empty set, we know the
+		// intersection is the empty set.
+		if len(s) == 0 {
+			return []T{}
+		}
+
+		m := make(map[T]struct{}, len(s))
+		for _, v := range s {
+			m[v] = struct{}{}
+		}
+		maps[i] = m
+	}
+
+	result := IntersectMapKeys(maps...)
+	final := make([]T, 0, len(result))
+	for k := range result {
+		final = append(final, k)
+	}
+	return final
+}
+
+// IntersectStable has the same invariants [Intersect], except it preserves the
+// relative order of elements in the intersection. This function has a
+// comparatively higher runtime complexity than [Intersect].
```

Comment:
Runtime could be maintained by just first running `IntersectMapKeys()` and then doing a linear pass through `slices` removing any key from the intersect map once seeing it.

Reply:
Psudedocode:

```
maps  = []
for slice in slices:
   maps.append(sliceToSet(slice))
intersectSet := IntersectMapKeys(maps)

final = [len(intersectSet)]
for slice in slices:
   for element in slice:
       if element in intersectSet:
             intersectSet.delete(element)
             final.append(element)

```

Reply:
I think it'll depend on the slice of the maps. For small maps (hundreds of elements), there's very little difference. Once you get into thousands or tens of thousands, this is comparatively more expensive. However, I would also argue that you shouldn't actually care about the order of 10k+ elements in a single slice.